### DISCUSSION TITLE
Accessing state from an async command (implicit elided lifetime not allowed here)

### DISCUSSION THREAD

**Discussion by alex-brook** [tauri-apps/tauri#4317] [Q&A] [answered]
Hi guys. I'm trying to manage a sqlite connection in my app, but I get a lifetime error when I try accessing it in the same way I would a non-async command.

```rust
  tauri::Builder::default()
    .manage(app_dir)
    .manage(Mutex::new(connection)) // Mutex<SqliteConnection>
    .invoke_handler(tauri::generate_handler!(create_todo))
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
```

```rust
#[tauri::command]
async fn create_todo(state: tauri::State<Mutex<SqliteConnection>>) {
  println!("Creating a todo");
  // Run insert query
}

// implicit elided lifetime not allowed here

// note: assuming a `'static` lifetime...
// help: indicate the anonymous lifetime: `tauri::State<'_, Mutex<SqliteConnection>>`rustc(E0726)
```

When I try and follow the compiler's suggestion I get a new error:

```rust
#[tauri::command]
async fn create_todo(state: tauri::State<'_, Mutex<SqliteConnection>>) {
  println!("Creating a todo");
  // Run insert query
}

// `__tauri_message__` does not live long enough

// borrowed value does not live long enoughrustc(E0597)
// main.rs(61, 1): borrowed value does not live long enough
// main.rs(61, 17): `__tauri_message__` dropped here while still borrowed
// `__tauri_message__` does not live long enough

// argument requires that `__tauri_message__` is borrowed for `'static`rustc(E0597)
// main.rs(61, 1): argument requires that `__tauri_message__` is borrowed for `'static`
// main.rs(61, 17): `__tauri_message__` dropped here while still borrowed
```

The first version compiles fine without the `async` keyword. What is going on here and what's the right way to do this?

**Comment by FabianLars** [comment, marked as answer]
https://github.com/tauri-apps/tauri/issues/2533

This is a known issue/limitation. You need to return a Result from async stateful commands. Something as simple as `Result<(), ()>` should be enough.

**Reply by alex-brook** [reply]
Thanks for the quick response! Can confirm this workaround works for me

**Reply by hwmaier** [reply]
> #2533
> 
> This is a known issue/limitation. You need to return a Result from async stateful commands. Something as simple as `Result<(), ()>` should be enough.

May I suggest to document this and also add it to the example here https://tauri.app/v1/guides/features/command#async-commands as this is not really obvious.
